gh-142349: Implement PEP 810 - Explicit lazy imports

pablogsal opened the pull request:
<!-- gh-issue-number: gh-142349 -->
* Issue: gh-142349
<!-- /gh-issue-number -->


⚠️ Dear reviewers, to avoid collapsing the GitHub API with a lot of comments, please open PRs against the base branch with any suggestions or fixes if you are sure are not controversial ⚠️ 

<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--142351.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

DinoV committed:
Lazy imports grammar / AST changes

DinoV committed:
Add __lazy_import__, check sys.modules before import

DinoV committed:
Export lazy_import in imp module

DinoV committed:
Add lazy import filter

DinoV committed:
Add compatiblity mode

DinoV committed:
Flow import func through to lazy imports object and __lazy_import__

DinoV committed:
Reify lazy objects when accessed via the module object

DinoV committed:
Import lazy.get

pablogsal committed:
Implement better error

pablogsal committed:
Fix reference and global dict in spezializing LOAD_GLOBAL

pablogsal committed:
Remove copyright

pablogsal committed:
More fixes

pablogsal committed:
Fix recursive lazy imports and error path in bytecodes.c

pablogsal committed:
fix offset in addr2line

pablogsal committed:
Add global flag

pablogsal committed:
Draft: force * imports to be eager

pablogsal committed:
Syntax restrictions for lazy imports

pablogsal committed:
Fix submodules crash

DinoV committed:
Implement disabling imports in try/except and * imports, report errors on bad usage of lazy

DinoV committed:
Add PyExc_ImportCycleError and raise it when a cycle is detected

DinoV committed:
Publish lazy imported packages on parent

DinoV committed:
Re-enable eagerly returning imported module

pablogsal committed:
Fix __lazy_modules__

pablogsal committed:
Add sys.set_lazy_imports_filter

pablogsal committed:
Simplify grammar

DinoV committed:
Move __lazy_imports__ check into the interpreter

pablogsal committed:
Make imports in with blocks syntax errors

DinoV committed:
Don't allow __lazy_imports__ to work in try/except

DinoV committed:
Move eager check for from imports into import from

johnslavik committed:
Highlight lazy imports in the new REPL

pablogsal committed:
Merge pull request #2 from bswck/lazy-import-pyrepl-highlight

Highlight lazy imports in the new REPL

pablogsal committed:
Fix global membership in LOAD_NAME

pablogsal committed:
C was a mistake

DinoV committed:
dir() doesn't reify module

pablogsal committed:
Merge pull request #7 from LazyImportsCabal/lazy_dir

dir() doesn't reify module

DinoV committed:
Add test case for external usage

DinoV committed:
Merge pull request #8 from LazyImportsCabal/lazy_test

Lazy test

DinoV committed:
Expose LazyImportType in types module

DinoV committed:
__import__ is loaded at reification time

DinoV committed:
Fix leaks and LOAD_ATTR specialization

pablogsal committed:
Merge pull request #9 from LazyImportsCabal/cleanup

More PEP-810 compatibility, fix memory leaks and LOAD_ATTR specialization

DinoV committed:
Update sys module to conform with the PEP, add matching C API

DinoV committed:
Merge pull request #10 from LazyImportsCabal/sys_c_api

Update sys module to conform with the PEP, add matching C API

DinoV committed:
Fix specialization of load global

pablogsal committed:
Merge pull request #12 from DinoV/lazy_mt

pablogsal committed:
Fix bug in specialization and make reification atomic

pablogsal committed:
Fix another race

DinoV committed:
Merge pull request #14 from LazyImportsCabal/fix

Fix bug in specialization and make reification atomic

DinoV committed:
Make __dict__ not reify to match the PEP

DinoV committed:
Allow try/except in with block

pablogsal committed:
Merge pull request #16 from LazyImportsCabal/lazy_globals

Lazy globals

DinoV committed:
Grab builtins dict from module

DinoV committed:
Merge pull request #17 from DinoV/lazy_builtins

Grab builtins dict from module

pablogsal committed:
Fix eager imports in try/except in global mode

DinoV committed:
Merge pull request #18 from LazyImportsCabal/bad_global

Fix eager imports in try/except in global mode

pablogsal committed:
Don't reify on REPL completion

DinoV committed:
Add tests for __lazy_import__

DinoV committed:
Merge remote-tracking branch 'origin/main' into HEAD

DinoV committed:
Merge pull request #21 from LazyImportsCabal/lazy_rebase

Lazy rebase

pablogsal committed:
Merge pull request #19 from DinoV/lazy_tests_dunder

DinoV committed:
Make sys.lazy_modules reflect the internal structure currently in use

pablogsal committed:
Implement more of PEP 810

pablogsal committed:
Add more tests

pablogsal committed:
Add more tests

pablogsal committed:
Merge remote-tracking branch 'upstream/main' into lazy

pablogsal committed:
Regen stuff

pablogsal committed:
Ah yes....windows

pablogsal committed:
More windows stuff and news

StanFromIreland commented on Doc/c-api/import.rst:
```diff
@@ -341,6 +341,43 @@ Importing Modules
+   - ``PyImport_LAZY_ALL``
+   - ``PyImport_LAZY_NONE``
+
+   .. versionadded:: 3.12
```
```suggestion
   .. versionadded:: next
```

ZeroIntensity commented on Doc/c-api/import.rst:
```diff
@@ -341,6 +341,43 @@ Importing Modules
+   Enumeration of possible lazy import modes:
+   - ``PyImport_LAZY_NORMAL``
+   - ``PyImport_LAZY_ALL``
+   - ``PyImport_LAZY_NONE``
```
Sphinx has `:c:enumerator:` for this.

StanFromIreland commented on Doc/whatsnew/3.15.rst:
```diff
@@ -77,6 +79,97 @@ Summary -- Release highlights
+used during a particular run.
+
+Developers have worked around this by moving imports inside functions, using
+``importlib`` to load modules on demand, or restructuring code to avoid
```
```suggestion
:mod:`importlib` to load modules on demand, or restructuring code to avoid
```

ZeroIntensity commented on Doc/c-api/import.rst:
```diff
@@ -341,6 +341,43 @@ Importing Modules
+
+.. c:function:: PyObject* PyImport_GetLazyImportsFilter()
+
+   Gets the current lazy imports filter. Returns a new reference.
```
Let's say "strong reference" instead of "new reference", since that's in the glossary. It'd also be good to clarify the return type.

StanFromIreland commented on Doc/whatsnew/3.15.rst:
```diff
@@ -77,6 +79,97 @@ Summary -- Release highlights
+discipline to apply consistently.
+
+Python now provides a cleaner solution through explicit lazy imports using the
+new ``lazy`` soft keyword. When you mark an import as lazy, Python defers the
```
Maybe add a link to it's doc here?

StanFromIreland commented on Grammar/python.gram:
```diff
@@ -121,9 +121,9 @@ simple_stmts[asdl_stmt_seq*]:
 simple_stmt[stmt_ty] (memo):
     | assignment
     | &"type" type_alias
+    | &('import' | 'from' | "lazy" ) import_stmt
```
```suggestion
    | &('import' | 'from' | "lazy") import_stmt
```

ZeroIntensity commented on Doc/c-api/import.rst:
```diff
@@ -341,6 +341,43 @@ Importing Modules
+   - ``PyImport_LAZY_ALL``
+   - ``PyImport_LAZY_NONE``
+
+   .. versionadded:: 3.12
```
3.12?
```suggestion
   .. versionadded:: 3.15
```

ZeroIntensity replied:
This looks resolved already.

ZeroIntensity commented on Doc/c-api/import.rst:
```diff
@@ -341,6 +341,43 @@ Importing Modules
+   Sets the current lazy imports filter. The function should be a callable that
+   will receive (importing_module_name, imported_module_name, [fromlist]) when
+   an import can potentially be lazy. Returns True if the import should be lazy
+   or False otherwise.
```
`True` and `False` are code names:
```suggestion
   an import can potentially be lazy. Returns ``True`` if the import should be lazy
   or ``False`` otherwise.
```

StanFromIreland commented on Objects/moduleobject.c:
```diff
@@ -50,6 +51,7 @@ assert_def_missing_or_redundant(PyModuleObject *m)
 }
 
 
+
```
```suggestion
```

pablogsal committed:
Updated versions

ZeroIntensity commented on Modules/_typesmodule.c:
You also need to update the pure-Python `types.py`.

pablogsal committed:
Update Doc/whatsnew/3.15.rst

Co-authored-by: Stan Ulbrych <89152624+StanFromIreland@users.noreply.github.com>

ZeroIntensity commented on Objects/lazyimportobject.c:
```diff
@@ -0,0 +1,186 @@
+    Py_INCREF(from);
+    m->lz_from = from;
+    Py_XINCREF(attr);
+    m->lz_attr = attr;
```
Wouldn't it be easier to use `Py_XNewRef`?
```suggestion
    m->lz_builtins = Py_XNewRef(builtins);
    m->lz_from = Py_XNewRef(from);
    m->lz_attr = Py_XNewRef(attr);
```

ZeroIntensity commented on Objects/lazyimportobject.c:
```diff
@@ -0,0 +1,186 @@
+        }
+    }
+
+    PyObject_GC_Track(m);
```
We usually use `_PyObject_GC_TRACK` and `_PyObject_GC_UNTRACK` internally.

ZeroIntensity commented on Objects/lazyimportobject.c:
```diff
@@ -0,0 +1,186 @@
+    Py_XDECREF(m->lz_builtins);
+    Py_XDECREF(m->lz_from);
+    Py_XDECREF(m->lz_attr);
+    Py_XDECREF(m->lz_code);
```
Up to you, but it's more concise to just call the `tp_clear` function in destructors.

ZeroIntensity commented on Objects/lazyimportobject.c:
```diff
@@ -0,0 +1,186 @@
+        }
+    }
+    Py_INCREF(m->lz_from);
+    return m->lz_from;
```
What's the intended behavior here? This will currently crash. If `m->lz_attr` is `NULL`, then the `return`s above aren't hit and this will attempt to increment `NULL`'s reference count.

ZeroIntensity commented on Python/bltinmodule.c:
```diff
@@ -287,6 +288,56 @@ builtin___import___impl(PyObject *module, PyObject *name, PyObject *globals,
+        locals = globals;
+    }
+
+    builtins = PyMapping_GetItemString(globals, "__builtins__");
```
I'm almost certain there's a `_Py_ID` for this.

StanFromIreland committed:
Add support in IDLE colorizer

pablogsal committed:
Merge pull request #22 from StanFromIreland/lazy-idle

pablogsal committed:
Address feedback

pablogsal committed:
Fix tests

pablogsal committed:
Ruff fixes

StanFromIreland commented on Doc/whatsnew/3.15.rst:
```diff
@@ -77,6 +79,97 @@ Summary -- Release highlights
+
+.. seealso:: :pep:`810` for the full specification and rationale.
+
+(Contributed by Pablo Galindo Salgado and Dino Viehland.)
```
```suggestion
(Contributed by Pablo Galindo Salgado and Dino Viehland in :gh:`142349`.)
```

pablogsal committed:
Update Doc/whatsnew/3.15.rst

Co-authored-by: Stan Ulbrych <89152624+StanFromIreland@users.noreply.github.com>

pablogsal committed:
JIT fixes

pablogsal committed:
More jit fixes

pablogsal committed:
Fix smelly

pablogsal committed:
Fix smelly

StanFromIreland committed:
Fix repl not coloring 'lazy' after 'from'

pablogsal committed:
Doc fixes

pablogsal committed:
brrrrrrrrrrrrrrrrrrrrrr

StanFromIreland commented on Lib/_pyrepl/utils.py:
```diff
@@ -276,6 +276,8 @@ def is_soft_keyword_used(*tokens: TI | None) -> bool:
             TI(T.NAME, string=s)
         ):
             return not keyword.iskeyword(s)
+        case (None | TI(T.NEWLINE) | TI(T.INDENT) | TI(T.DEDENT), TI(string="lazy"), TI(string="import")):
```
This does not account for `lazy from ...`, I have the suggestion here: https://github.com/LazyImportsCabal/cpython/pull/24

pablogsal committed:
Merge pull request #24 from StanFromIreland/lazy-color

pablogsal committed:
PeRfoRManCe

pablogsal committed:
PeRfoRManCe

picnixz committed:
Amend return types for `PyImport_SetLazy*` functions

picnixz commented on Doc/using/cmdline.rst:
```diff
@@ -694,6 +694,14 @@ Miscellaneous options
+     (the default) respects the ``lazy`` keyword in source code.
+     See also :envvar:`PYTHON_LAZY_IMPORTS`.
+
+     .. versionadded:: 3.15
```
```suggestion
     .. versionadded:: next
```

picnixz commented on Doc/library/types.rst:
```diff
@@ -343,6 +343,18 @@ Standard names are defined for the following types:
+   actually accessed. This type can be used to detect lazy imports
+   programmatically.
+
+   .. versionadded:: 3.15
```
```suggestion
   .. versionadded:: next
```

picnixz commented on Include/import.h:
```diff
@@ -88,6 +88,18 @@ PyAPI_FUNC(int) PyImport_AppendInittab(
     PyObject* (*initfunc)(void)
     );
 
+typedef enum  {
```
```suggestion
typedef enum {
```

picnixz commented on Include/internal/pycore_magic_number.h:
```diff
@@ -287,6 +287,7 @@ Known values:
     Python 3.15a1 3654 (Fix missing exception handlers in logical expression)
     Python 3.15a1 3655 (Fix miscompilation of some module-level annotations)
     Python 3.15a1 3656 (Add TRACE_RECORD instruction, for platforms with switch based interpreter)
+    Python 3.15a1 3657 Lazy imports IMPORT_NAME opcode changes
```
```suggestion
    Python 3.15a3 3657 Lazy imports IMPORT_NAME opcode changes
```

I _think_ we already released 3.15a2 right?

pablogsal committed:
moar fixes much refleaks wow

picnixz commented on Include/internal/pycore_magic_number.h:
```diff
@@ -300,7 +301,7 @@ PC/launcher.c must also be updated.
 */
 
-#define PYC_MAGIC_NUMBER 3656
+#define PYC_MAGIC_NUMBER 3658
```
Wgy are we bumping it by two?

pablogsal replied:
Because a bad 3rd rebase against main :sweat:

pablogsal committed:
Merge pull request #25 from picnixz/patch-1

picnixz commented on Lib/importlib/_bootstrap.py:
```diff
@@ -1359,6 +1359,12 @@ def _find_and_load_unlocked(name, import_):
+    # Set attributes to lazy submodules on the module.
+    try:
+        _imp._set_lazy_attributes(module, name)
+    except Exception as e:
```
Don't you need to possible clear `e` to avoid a possible reference cycle here?

pablogsal replied:
IIRC is cleared automatically

pablogsal replied:
```
main lazy  ? ❯ cat lel.py
def foo():
    try:
        1/0
    except Exception as e:
        print(e)

    pass

main lazy  ? ❯ ./python -m dis lel.py
  0           RESUME                   0

  1           LOAD_CONST               0 (<code object foo at 0x7f3d9cd668f0, file "lel.py", line 1>)
              MAKE_FUNCTION
              STORE_NAME               0 (foo)
              LOAD_CONST               1 (None)
              RETURN_VALUE

Disassembly of <code object foo at 0x7f3d9cd668f0, file "lel.py", line 1>:
   1            RESUME                   0

   2            NOP

   3    L1:     LOAD_SMALL_INT           1
                LOAD_SMALL_INT           0
                BINARY_OP               11 (/)
                POP_TOP

   7    L2:     LOAD_CONST               1 (None)
                RETURN_VALUE

  --    L3:     PUSH_EXC_INFO

   4            LOAD_GLOBAL              0 (Exception)
                CHECK_EXC_MATCH
                POP_JUMP_IF_FALSE       23 (to L9)
        L4:     NOT_TAKEN
        L5:     STORE_FAST               0 (e)

   5    L6:     LOAD_GLOBAL              3 (print + NULL)
                LOAD_FAST                0 (e)
                CALL                     1
                POP_TOP
        L7:     POP_EXCEPT
                LOAD_CONST               1 (None)
                STORE_FAST               0 (e)
                DELETE_FAST              0 (e)

   7            LOAD_CONST               1 (None)
                RETURN_VALUE

  --    L8:     LOAD_CONST               1 (None)
                STORE_FAST               0 (e)
                DELETE_FAST              0 (e)
                RERAISE                  1

   4    L9:     RERAISE                  0

  --   L10:     COPY                     3
                POP_EXCEPT
                RERAISE                  1
ExceptionTable:
  L1 to L2 -> L3 [0]
  L3 to L4 -> L10 [1] lasti
  L5 to L6 -> L10 [1] lasti
  L6 to L7 -> L8 [1] lasti
  L8 to L10 -> L10 [1] lasti
```

picnixz commented on Lib/rlcompleter.py:
```diff
@@ -188,7 +189,16 @@ def attr_matches(self, text):
+
+                    if (isinstance(thisobject, types.ModuleType)
+                        and
+                        isinstance(thisobject.__dict__.get(word),types.LazyImportType)
```
```suggestion
                        isinstance(thisobject.__dict__.get(word), types.LazyImportType)
```

pablogsal committed:
Doc updates

pablogsal committed:
Address feedback

picnixz commented on Python/import.c:
```diff
@@ -4304,6 +4743,52 @@ PyImport_ImportModuleAttrString(const char *modname, const char *attrname)
+PyImport_GetLazyImportsFilter(void)
+{
+    PyInterpreterState *interp = _PyInterpreterState_GET();
+    return Py_XNewRef(FT_ATOMIC_LOAD_PTR_RELAXED(LAZY_IMPORTS_FILTER(interp)));
```
If the import filter is not set, this may return `NULL` without an exception set. So either you want it to be documented or you want to return `None` instead.

picnixz commented on Python/import.c:
```diff
@@ -4304,6 +4743,52 @@ PyImport_ImportModuleAttrString(const char *modname, const char *attrname)
+    PyObject *old = _Py_atomic_exchange_ptr(&LAZY_IMPORTS_FILTER(interp), Py_XNewRef(filter));
+    Py_XDECREF(old);
+#else
+    Py_XSETREF(LAZY_IMPORTS_FILTER(interp), Py_XNewRef(filter));
```
This makes me wondeer whether we couldn't have a FT helper for this. It's probably not the only place that will need this in the future.

picnixz commented on Python/import.c:
```diff
@@ -3963,6 +4211,193 @@ PyImport_ImportModuleLevelObject(PyObject *name, PyObject *globals,
+    PyInterpreterState *interp = tstate->interp;
+    assert(_PyEval_GetFrame()->f_globals == _PyEval_GetFrame()->f_locals); // should only be called in global scope
+
+    // Check if the filter disables the lazy import
+    PyObject *filter = FT_ATOMIC_LOAD_PTR_RELAXED(LAZY_IMPORTS_FILTER(interp));
+    if (filter != NULL) {
+        PyObject *modname;
+        if (PyDict_GetItemRef(globals, &_Py_ID(__name__), &modname) < 0) {
+            return NULL;
+        } else if (modname == NULL) {
+            modname = Py_None;
+        }
```
This should be two separate `if`s right? because you might have `modname == NULL` if the name doesn't exist.

picnixz commented on Python/import.c:
```diff
@@ -3963,6 +4211,193 @@ PyImport_ImportModuleLevelObject(PyObject *name, PyObject *globals,
+    if (filter != NULL) {
+        PyObject *modname;
+        if (PyDict_GetItemRef(globals, &_Py_ID(__name__), &modname) < 0) {
+            return NULL;
```
You need to decref `abs_name` here (you can also have a goto error maybe?)

pablogsal committed:
More doc fixes

picnixz commented on Python/import.c:
```diff
@@ -3963,6 +4211,193 @@ PyImport_ImportModuleLevelObject(PyObject *name, PyObject *globals,
+
+        if (!PyObject_IsTrue(res)) {
+            Py_DECREF(abs_name);
+            return PyImport_ImportModuleLevelObject(
```
Just me thinking aloud but is it possible to have a UAF here because `filter` is an arbitrary callable (so, we could alter `fromlist`)?

picnixz commented on Python/import.c:
```diff
@@ -3963,6 +4211,193 @@ PyImport_ImportModuleLevelObject(PyObject *name, PyObject *globals,
+
+    // Add the module name to sys.lazy_modules set (PEP 810)
+    PyObject *lazy_modules_set = interp->imports.lazy_modules_set;
+    if (lazy_modules_set != NULL) {
```
When is it possible to have `lazy_modules_set` NULL?

picnixz commented on Python/import.c:
```diff
@@ -4903,6 +5388,66 @@ _imp_source_hash_impl(PyObject *module, long key, Py_buffer *source)
+        }
+
+        child_dict = get_mod_dict(child_module);
+        if (child_dict == NULL || !PyDict_CheckExact(child_dict)) {
```
child_dict can be NULL and we must go to error here

StanFromIreland committed:
Make error for invalid `lazy from __future__ ...` import prettier

pablogsal committed:
Regen limited ABI

StanFromIreland committed:
Merge branch 'lazy' into lazy-future-error

StanFromIreland commented on Doc/reference/simple_stmts.rst:
```diff
@@ -748,14 +748,15 @@ The :keyword:`!import` statement
    pair: name; binding
    pair: keyword; from
    pair: keyword; as
+   pair: keyword; lazy
```
I suggest you move this to the below section on lazy specifically.

picnixz reviewed (commented):
I'll review more tomorrow when I'm on my dev session, it'll be easier. But here are a few comments and questions about `import.c`

StanFromIreland commented on Parser/action_helpers.c:
```diff
@@ -1943,17 +1943,21 @@ _PyPegen_concatenate_strings(Parser *p, asdl_expr_seq *strings,
                       		   PyArena *arena) {
     if (level == 0 && PyUnicode_CompareWithASCIIString(module, "__future__") == 0) {
+        if (is_lazy) {
+            RAISE_SYNTAX_ERROR("lazy from __future__ import is not allowed");
```
To improve the location info, I propose a little change in https://github.com/LazyImportsCabal/cpython/pull/26

pablogsal replied:
Thanks a lot!

pablogsal committed:
Fix error paths and edge cases

StanFromIreland commented on Doc/reference/simple_stmts.rst:
```diff
@@ -870,6 +871,56 @@ determine dynamically the modules to be loaded.
+names remain as lazy proxies until they are accessed.
+
+Example::
+
+   lazy import json
```
```suggestion

   lazy import json
   lazy import sys
```

For `'json' in sys.modules` to run.

pablogsal committed:
Free threading fixes

picnixz commented on Include/internal/pycore_magic_number.h:
```diff
@@ -300,7 +301,7 @@ PC/launcher.c must also be updated.
 */
 
-#define PYC_MAGIC_NUMBER 3656
+#define PYC_MAGIC_NUMBER 3658
```
Sorry but I think you forgot to update this magic number

pablogsal committed:
TSAN fixes

pablogsal committed:
Update magic number

pablogsal committed:
Merge pull request #26 from StanFromIreland/lazy-future-error

pablogsal committed:
Regen frozenmain

pablogsal committed:
Fix WASI failure

StanFromIreland commented on Lib/test/test_import/test_lazy_imports.py:
```diff
@@ -0,0 +1,1535 @@
+    def test_global_off(self):
+        """Mode 'none' should disable lazy imports entirely."""
+        import test.test_import.data.lazy_imports.global_off
+        self.assertTrue("test.test_import.data.lazy_imports.basic2" in sys.modules)
```
We can use `assertIn` instead, a suggestion that applies for all cases: https://github.com/LazyImportsCabal/cpython/pull/27

StanFromIreland commented on Lib/test/.ruff.toml:
```diff
@@ -13,6 +13,9 @@ extend-exclude = [
     # New grammar constructions may not yet be recognized by Ruff,
     # and tests re-use the same names as only the grammar is being checked.
     "test_grammar.py",
+    # Lazy import syntax (PEP 810) not yet supported by Ruff
```
```suggestion
    # Lazy import syntax (PEP 810) is not yet supported by Ruff
```

picnixz committed:
fix error path in `_PyImport_LoadLazyImportTstate`

picnixz committed:
fix UAFs in `register_lazy_on_parent`

picnixz committed:
fix error path in `_PyImport_LazyImportModuleLevelObject`

picnixz committed:
fix error path in `_imp__set_lazy_attributes_impl`

picnixz committed:
document return value for `PyImport_GetLazyImportsFilter`

picnixz committed:
Various UAFs & cosmetic fixes in `Python/{bltinmodule,bytecode,ceval}.c` (#33)

* fix UAF in `bltinmodule.c`

* fix UAF in `bytecode.c`

* run `make regen-all`

* various improvements to `ceval.c`

* fix CI

picnixz committed:
Fix missing `lazy`-related symbols for the JIT (#34)

* fix missing `PyLazyImport_CheckExact`

* fix missing lazy-related symbols

* fix missing lazy-related symbols

* `extern` seems enough

* Export _PyImport_LoadLazyImportTstate for JIT

Yhg1s committed:
Merge branch 'main' into lazy

picnixz committed:
Fix and improve `Objects/lazyimportobject.c` (#31)

* re-order imports

* simplify static type definition

* fix undefined behaviors

* reject keyword arguments for `lazy_import`

Yhg1s committed:
Fix doc merge snafu.

Yhg1s committed:
Fix examples in the `ast` module docs (run by doctest).

Yhg1s committed:
Don't include the new C API functions
(`Py_(Get|Set)LazyImports(Filter|Mode)`) in the limited API (for now).

Yhg1s committed:
Add the new builtin types to the known-to-be-mutable-global lists in
Tools/c-analyzer.

pablogsal committed:
Fix some crashes

```
import types
types.LazyImportType({}, "3", 0)
```

```
def f():
    exec("lazy import json")
f()
```

Yhg1s committed:
Document ImportCycleError and PyExc_ImportCycleError.

StanFromIreland committed:
Use `self.assertIn/assertNotIn` instead of `self.assertTrue/assertFalse` (#27)

* Use self.assertIn and self.assertNotIn instead

* reorder imports

* use assertIn/assertNotIn/assertHasAttr

---------

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

StanFromIreland committed:
Fix more cases where lazy is not highlighted in repl (#36)

DinoV committed:
More thread safety fixes

DinoV committed:
_PyLazyImport_New improvements

Sacul0457 commented on Doc/whatsnew/3.15.rst:
```diff
@@ -77,6 +79,98 @@ Summary -- Release highlights
+
+   print("Starting up...")  # json and datetime not loaded yet
+
+   data = json.loads('{"key": "value"}')  # json loads here
```
```suggestion
   data = json.loads('{"key": "value"}')  # json gets loaded here
```
The comment `# json loads here` can be mis-leading where people might think it refers to calling `json.loads` and not loading the module. It might be better to maybe use a different example

DinoV committed:
Add lazy import dis tests

DinoV committed:
Add C API tests

DinoV committed:
Make comment clearer

DinoV committed:
Merge branch 'main' of github.com:python/cpython into lazy_head

DinoV committed:
Fixup clinic issue

DinoV committed:
Remove tsan fixes pending https://github.com/colesbury/cpython/tree/gh-142534-consume

DinoV committed:
Merge branch 'main' of github.com:python/cpython into HEAD

brittanyrey committed:
Standardize versionadded tags and minor updates to documentation

Yhg1s committed:
Combine two very similar sentences together.

brittanyrey committed:
Address open comments
* Update example to add missing import
* Add link to lazy keyword
* Update documentation

picnixz committed:
Apply PEP-7 to `Python/import.c` (#29)

* PEP-7 for import.c

* re-order imports

* fix tests

* some final nits

* alter other file

* alter sysmodule.c

Yhg1s committed:
Reflow and reformat to stick to PEP 7's line length limit in most places,
and prefer using `//` for comments instead of `/* */`.

pablogsal committed:
Merge branch 'main' into lazy

pablogsal committed:
Merge pull request #38 from Yhg1s/reflow-reformat

hugovk commented on Doc/whatsnew/3.15.rst:
```diff
@@ -80,6 +82,105 @@ Summary -- Release highlights
+accept three values: ``all`` makes all imports lazy by default, ``none``
+disables lazy imports entirely (even explicit ``lazy`` statements become
+eager), and ``normal`` (the default) respects the ``lazy`` keyword in source
+code. The :func:`sys.set_lazy_imports` and :func:`sys.get_lazy_imports`
```
There's a `sys.lazy_modules` dict but these functions don't exist yet.

pablogsal replied:
@brittanyrey do you want to give this a go?

brittanyrey replied:
Just to confirm @hugovk which part is missing?
I'm seeing:
```
>>> import sys
>>> sys.lazy_modules
{}
>>> sys.get_lazy_imports()
'normal'
>>> sys.set_lazy_imports("all")
>>> sys.get_lazy_imports()
'all'
```

hugovk replied:
Ah sorry, I'm on another machine now but the other one must have still been on an old branch. Thanks, looks good!

hugovk replied:
Back on this machine now. The branch wasn't old, but I meant to comment on `get_lazy_modules`, not `get_lazy_imports`.

What's new says:

```rst
For debugging and introspection, :func:`sys.get_lazy_modules` returns a set
containing the names of all modules that have been lazily imported but not
yet loaded.
```
And `sys.rst`:

```rst
.. function:: get_lazy_modules()

   Returns a set of fully-qualified module names that have been
   lazily imported. This is primarily useful for diagnostics and
   introspection.

   Note that modules are removed from this set when they are reified
   (actually loaded on first use).
```

And this function doesn't exist:

```pytb
Python 3.15.0a3+ (heads/lazy:49712b7ce9c, Dec 18 2025, 12:56:10) [Clang 17.0.0 (clang-1700.3.19.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> sys.get_lazy_modules()
Traceback (most recent call last):
  File "<python-input-1>", line 1, in <module>
    sys.get_lazy_modules()
    ^^^^^^^^^^^^^^^^^^^^
AttributeError: module 'sys' has no attribute 'get_lazy_modules'. Did you mean: 'lazy_modules'?
```

```console
❯ rg get_lazy_modules -l
Doc/whatsnew/3.15.rst
Doc/library/sys.rst
```

The [PEP](https://peps.python.org/pep-0810/) doesn't mention this function. Should it be implemented, or the docs removed?

Yhg1s replied:
Yes, that's a reference I think Dino forgot to remove in 4fd8fe8a833321374840d199f43acd5a32d1063d. There's just `sys.lazy_modules` now.

hugovk replied:
Removed from docs in https://github.com/LazyImportsCabal/cpython/pull/40 👍

pablogsal committed:
Do not reify in tracebacks

hugovk commented on Doc/library/sys.rst:
```diff
@@ -911,6 +911,47 @@ always available. Unless explicitly noted otherwise, all variables are read-only
+.. function:: get_lazy_modules()
+
+   Returns a set of fully-qualified module names that have been
+   lazilyimported. This is primarily useful for diagnostics and
```
```suggestion
   lazily imported. This is primarily useful for diagnostics and
```

pablogsal committed:
Merge branch 'main' into lazy

johnslavik committed:
Fix comment about `check_lazy_import_compatibility` (#39)

pablogsal committed:
Update Doc/library/sys.rst

Co-authored-by: Hugo van Kemenade <1324225+hugovk@users.noreply.github.com>

Yhg1s committed:
Regen generated files.

Yhg1s committed:
Merge branch 'main' into lazy

JelleZijlstra commented on Doc/c-api/import.rst:
```diff
@@ -346,6 +346,58 @@ Importing Modules
+
+.. c:function:: int PyImport_SetLazyImportsMode(PyImport_LazyImportsMode mode)
+
+   Similar to :c:func:`PyImport_ImportModuleAttr`, but names are UTF-8 encoded
```
Copy-paste error? This seems unrelated to the function being described.

JelleZijlstra commented on Doc/reference/simple_stmts.rst:
```diff
@@ -870,6 +871,57 @@ determine dynamically the modules to be loaded.
+
+   print('json' in sys.modules)  # True - now loaded
+
+If an error occurs during module loading (such as :exc:`ImportError` or
```
Perhaps also document that ImportCycleError may be raised?

JelleZijlstra commented on Lib/test/test_dis.py:
```diff
@@ -292,7 +292,7 @@ def wrap_func_w_kwargs():
   1           LOAD_SMALL_INT           0
               LOAD_CONST               1 (('*',))
-              IMPORT_NAME              0 (math)
+              IMPORT_NAME              2 (math + eager)
```
Consider adding more tests for the new logic in dis.py.

JelleZijlstra commented on Lib/test/test_syntax.py:
```diff
@@ -3475,6 +3475,119 @@ def test_ifexp_body_stmt_else_stmt(self):
+try:
+    lazy import os
+except:
+    pass
```
What about putting it in the `except:`?

brittanyrey replied:
Added a [commit](https://github.com/python/cpython/pull/142351/commits/9fc7da23a6e74bab73bca608ba929f8d1e2d4668) to address

kumaraditya303 commented on Include/internal/pycore_dict.h:
```diff
@@ -36,6 +36,16 @@ extern int _PyDict_DelItem_KnownHash_LockHeld(PyObject *mp, PyObject *key,
 extern int _PyDict_Contains_KnownHash(PyObject *, PyObject *, Py_hash_t);
 
+// "Id" variants
+extern PyObject* _PyDict_GetItemIdWithError(PyObject *dp,
```
Looks like this is from a bad merge? The Id variants were removed from main recently and were not allowed to be used in internal code for a while now.

Yhg1s replied:
Yep, fixed.

kumaraditya303 commented on Objects/lazyimportobject.c:
```diff
@@ -0,0 +1,180 @@
+        PyErr_SetString(PyExc_TypeError, "expected str for name");
+        return NULL;
+    }
+    if (fromlist == Py_None || fromlist == NULL) {
```
```suggestion
    if (fromlist == Py_None) {
```

Yhg1s replied:
No, fromlist can be NULL (see `attr` in `lazy_import_new`; admittedly the variable name mismatch is confusing there.)

kumaraditya303 replied:
On the next line it is assigned to `NULL` so the check is redundant.

Yhg1s replied:
There's an 'else' attached to the 'if', so it isn't redundant.

kumaraditya303 replied:
Oh, sorry, I missed that

hugovk committed:
Remove get_lazy_modules from docs

kumaraditya303 commented on Objects/lazyimportobject.c:
```diff
@@ -0,0 +1,180 @@
+    .tp_basicsize = sizeof(PyLazyImportObject),
+    .tp_dealloc = lazy_import_dealloc,
+    .tp_repr = lazy_import_repr,
+    .tp_flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_HAVE_GC | Py_TPFLAGS_BASETYPE,
```
It is noted in the docstring that this type is used internally. In that case does it need to be a base type? 

Also if `users should not typically create instances directly` then why not enforce that with `Py_TPFLAGS_DISALLOW_INSTANTIATION`?

DinoV replied:
I think it shouldn't be a base type.

We do allow people to create them now as the type is exposed via `types` so we should probably fix the doc string.

DinoV replied:
Fixed with [4d1129b](https://github.com/python/cpython/pull/142351/commits/4d1129bd0297369ecbbcb12175839be49f76ce09)

kumaraditya303 commented on Objects/moduleobject.c:
```diff
@@ -177,6 +178,7 @@ new_module_notrack(PyTypeObject *mt)
     m->md_state = NULL;
     m->md_weaklist = NULL;
     m->md_name = NULL;
+    m->m_dict_version = 0;
```
All other fields start with `md_`, can this use that as well?

Yhg1s replied:
I'm not even sure where this is being used right now... @DinoV, @pablogsal?

DinoV replied:
I think this got added when we reified the whole dict object on access, we should remove it.

DinoV replied:
Fixed with [4d1129b](https://github.com/python/cpython/pull/142351/commits/4d1129bd0297369ecbbcb12175839be49f76ce09)

Yhg1s committed:
Get rid of a merge snafu.

DinoV committed:
Remove cruft from module object left over from reification of __dict__

DinoV committed:
Fixup lazy import object, don't allow subclassing, improve doc, and variable names in lazy_import_new

DinoV committed:
Merge pull request #40 from hugovk/3.15-lazy-rm-get_lazy_modules

Remove `get_lazy_modules` from docs

brittanyrey committed:
Add test exercising that the appropriate error is raised when calling
lazy import within an except block. Also removed unused var.

javabster commented on Doc/reference/simple_stmts.rst:
```diff
@@ -870,6 +871,57 @@ determine dynamically the modules to be loaded.
+When an import statement is preceded by the :keyword:`lazy` keyword,
+the import becomes *lazy*: the module is not loaded immediately at the import
+statement. Instead, a lazy proxy object is created and bound to the name. The
+actual module is loaded on first use of that name.
```
I would recommend swapping the first two paragraphs around, the first sentence "the lazy keyword marks an import as lazy" is not super helpful if you don't define what you mean by "lazy" until the next paragraph 😅

```suggestion
When an import statement is preceded by the :keyword:`lazy` keyword,
the import becomes *lazy*: the module is not loaded immediately at the import
statement. Instead, a lazy proxy object is created and bound to the name. The
actual module is loaded on first use of that name. 

The :keyword:`lazy` keyword is a :ref:`soft keyword
<soft-keywords>` that only has special meaning when it appears immediately
before an :keyword:`import` or :keyword:`from` statement.
```

javabster replied:
Alternatively, you could keep the same order of the paragraphs but just remove the first sentence

Yhg1s replied:
I've removed the first sentence.

javabster commented on Doc/whatsnew/3.15.rst:
```diff
@@ -80,6 +82,102 @@ Summary -- Release highlights
+Python now provides a cleaner solution through explicit :keyword:`lazy`
+imports using the new ``lazy`` soft keyword. When you mark an import as
+lazy, Python defers the actual module loading until the imported name is
+first used. This gives you the organizational benefits of declaring all
+imports at the top of the file while only paying the loading cost for
+modules you actually use.
```
I think this is a good succinct point explaining why folks would want to use lazy imports - might be worth also mentioning this reasoning in the reference section

Yhg1s replied:
This doesn't really fit in the reference docs, which mostly try to explain the 'what' and 'how' without editorializing into 'why' (mostly, because it's not that uniform). Maybe it would make sense to go deeper into the 'why' somewhere other than the PEP, but that's how most things in the docs are laid out right now.

Yhg1s committed:
Merge branch 'main' into lazy

Yhg1s committed:
Fix test_sys after the removal of m_dict_version from the module object in
931d1c1ea504d2426087c2b50ef99a3f2f666f49.

Yhg1s committed:
Address Abby's docs comment.

johnslavik committed:
Remove `types.LazyImportType` from pure variant of `types` (#48)

Yhg1s committed:
Merge branch 'main' into lazy

markshannon commented on Objects/dictobject.c:
```diff
@@ -4674,6 +4674,14 @@ _PyDict_SizeOf_LockHeld(PyDictObject *mp)
 }
 
+void
+_PyDict_ClearKeysVersion(PyObject *mp)
```
```suggestion
_PyDict_ClearKeysVersionLockHeld(PyObject *mp)
```

Most functions that require a lock to be held indicate so in their name

markshannon commented on Objects/lazyimportobject.c:
```diff
@@ -0,0 +1,177 @@
+    m->lz_code = NULL;
+    m->lz_instr_offset = -1;
+
+    _PyInterpreterFrame *frame = _PyEval_GetFrame();
```
Why get the frame from the context in this non-obvious way?
Why not pass it in as a parameter like the builtins and name?

DinoV replied:
I think part of this was because we allowed arbitrary construction... I'll flow it down from the callers now.

markshannon commented on Objects/lazyimportobject.c:
```diff
@@ -0,0 +1,177 @@
+    .tp_clear = lazy_import_clear,
+    .tp_methods = lazy_import_methods,
+    .tp_alloc = PyType_GenericAlloc,
+    .tp_new = lazy_import_new,
```
I don't think this is a good idea. It isn't in the PEP and I can see no use for it.

markshannon commented on Objects/moduleobject.c:
```diff
@@ -1265,13 +1266,52 @@ _PyModule_IsPossiblyShadowing(PyObject *origin)
+    int err;
+    Py_BEGIN_CRITICAL_SECTION(dict);
+
+    _PyDict_ClearKeysVersion(dict);
```
`_PyModule_ReplaceLazyValue` doesn't change the keys, it changes the values, so the name "keys version" and associated documentation/comments are now wrong.
At the very least some renaming is needed.

DinoV replied:
We are just updating the value, but it is the dictionary keys version that we need to clear to invalidate the existing inline caches. I'm open to a better name but it seems like it might get really unwieldy if it includes the "ClearKeysVersionBecauseWeModifyValues"

markshannon commented on Python/bytecodes.c:
```diff
@@ -1827,7 +1857,21 @@ dummy_func(
+
             ERROR_IF(PyStackRef_IsNull(*res));
+
+            PyObject *res_o = PyStackRef_AsPyObjectBorrow(*res);
```
Can you move this into `_PyEval_LoadGlobalStackRef` which does the work of `_LOAD_GLOBAL` otherwise this bloats the JIT stencils

markshannon commented on Python/jit.c:
```diff
@@ -11,9 +11,11 @@
 #include "pycore_floatobject.h"
 #include "pycore_frame.h"
 #include "pycore_function.h"
+#include "pycore_import.h"
```
See my earlier comment about moving code into the helper function.

markshannon reviewed (changes requested):
I've a few comments.

More importantly, is it supposed to work this way?
```py
lazy import module  # module is just one = 1

class C: pass
sneaky = C()
sneaky.one = 1

def f():
    t = 0
    for _ in range(10):
        t += sneaky.one
    return t

print(f())
globals()["sneaky"] = globals()["module"]
print(f())
```
```
10
Traceback (most recent call last):
  File "test.py", line 17, in <module>
    print(f())
          ~^^
  File "test.py", line 12, in f
    t += sneaky.one
         ^^^^^^^^^^
AttributeError: 'lazy_import' object has no attribute 'one'
```

DinoV committed:
Merge branch 'main' of github.com:python/cpython into lazy_review_updates

DinoV committed:
Fix ups from code reviews

DinoV committed:
Use dictionary watchers to update version number on writes to dictionary

DinoV committed:
Small fixups

DinoV committed:
Merge pull request #49 from DinoV/lazy_review_updates

Lazy review updates

DinoV committed:
Flow _PyInterpreterFrame into _PyLazyImport_New

DinoV committed:
Merge branch 'main' of github.com:python/cpython into lazy_review_updates2

DinoV committed:
Fix test/test_types

DinoV committed:
Add include for jit stencils

DinoV committed:
Fix free threaded build

DinoV committed:
Merge pull request #50 from DinoV/lazy_review_updates2

Flow in _PyInterpreterFrame and fix issue when global not defined

brittanyrey committed:
Merge branch 'main' into lazy

DinoV committed:
Fix bad merge

DinoV committed:
Merge pull request #51 from DinoV/fix_merge

Fix bad merge

DinoV committed:
Merge branch 'main' of github.com:python/cpython into HEAD

DinoV reviewed (approved):
Woot! Woot!

Yhg1s committed:
Merge pull request #52 from DinoV/lazy_merge

Merge DinoV's merge of main.

Yhg1s committed:
Merge branch 'main' into lazy
